lib/charms/parca_k8s/v0/parca_scrape.py

```diff
@@ -705,143 +642,68 @@ def __init__(
-        Juju unit relation data.
-        """
-        self._publish_all_relation_data()
+        for relation in relations:
+            relation.data[app]["scrape_metadata"] = json.dumps(scrape_metadata)
+            relation.data[app]["scrape_jobs"] = json.dumps(jobs)
 
-    def _set_unit_ip(self, _event=None):
+    @staticmethod
+    def _publish_unit_data(unit: ops.Unit, relations: List[ops.Relation]):
         """Set unit host address.
 
         Each time a profiling provider charm container is restarted it updates its own host address
         in the unit relation data for the Parca charm. The only argument specified is an event and
         it is ignored.
         """
-        for relation in self._charm.model.relations[self._relation_name]:
-            relation.data[self._charm.unit]["parca_scrape_unit_address"] = socket.getfqdn()
-            relation.data[self._charm.unit]["parca_scrape_unit_name"] = str(
-                self._charm.model.unit.name
-            )
+        for relation in relations:
+            relation.data[unit]["parca_scrape_unit_address"] = socket.getfqdn()
+            relation.data[unit]["parca_scrape_unit_name"] = str(unit.name)
 
-    def _is_valid_unit_address(self, address: str) -> bool:
-        """Validate a unit address.
-
-        Args:
-            address: a string representing a unit address
-        """
-        try:
-            _ = ipaddress.ip_address(address)
-            return True
-        except ValueError:
-            return False
-
-    @property
-    def _relations(self) -> List[Relation]:
-        """The relations currently active on this endpoint."""
-        return self._charm.model.relations[self._relation_name]
-
-    def is_ready(self, relation: Optional[Relation] = None) -> bool:
+    def is_ready(self, charm:ops.CharmBase) -> bool:
         """Check if the relation(s) on this endpoint are ready."""
-        relations = [relation] if relation else self._relations
-
+        relations = charm.model.relations[self._relation_name]
         if not relations:
             logger.debug(f"no relation on {self._relation_name!r}.")
             return False
 
         # TODO: once we have a pydantic model, we can also check for the integrity of the databags.
-        return all((relation.app and relation.data) for relation in relations)
-
-    @property
-    def _scrape_jobs(self) -> list:
-        """Fetch list of scrape jobs.
-
-        Returns:
-           A list of dictionaries, where each dictionary specifies a single scrape job for Parca.
-        """
-        return self._jobs if self._jobs else [DEFAULT_JOB]
-
-    @property
-    def _scrape_metadata(self) -> dict:
-        """Generate scrape metadata.
-
-        Returns:
-            Scrape configuration metadata for this profiling provider charm.
-        """
-        return self.topology.as_dict()
+        return all((relation.app and relation.data) for relation in relations)
```

Comment:
Mesh had a debate today about the value of an `is_ready` in a provider (eg: in something that sends data).  I'm torn. 

Paraphrasing @dstathis, he wondered what value it adds.  He has a fair point too - if we've called `reconcile()` then `is_ready` should be redundant as either we successfully sent data or something bad enough went wrong that we should probably be in Error

On the other hand, `is_ready` is perfect for a `collect-unit-status` handler.  imo, its important to put the status of all the relations in `collect-unit-status`, because an application shouldnt be Active/Idle if it had relations it hasn't sent data to yet.

wdyt?